arrow-integrations-jackson-module/src/main/kotlin/arrow/integrations/jackson/module/OptionModule.kt

```diff
@@ -2,143 +2,36 @@ package arrow.integrations.jackson.module
 
 import arrow.core.None
 import arrow.core.Option
-import arrow.core.getOrElse
-import arrow.core.or
-import com.fasterxml.jackson.core.JsonParser
+import arrow.core.some
 import com.fasterxml.jackson.core.json.PackageVersion
-import com.fasterxml.jackson.databind.BeanDescription
-import com.fasterxml.jackson.databind.BeanProperty
-import com.fasterxml.jackson.databind.DeserializationContext
-import com.fasterxml.jackson.databind.JavaType
-import com.fasterxml.jackson.databind.JsonDeserializer
-import com.fasterxml.jackson.databind.JsonSerializer
-import com.fasterxml.jackson.databind.MapperFeature
-import com.fasterxml.jackson.databind.SerializationConfig
-import com.fasterxml.jackson.databind.deser.ContextualDeserializer
-import com.fasterxml.jackson.databind.jsontype.TypeSerializer
 import com.fasterxml.jackson.databind.module.SimpleModule
-import com.fasterxml.jackson.databind.ser.Serializers
-import com.fasterxml.jackson.databind.ser.std.ReferenceTypeSerializer
-import com.fasterxml.jackson.databind.type.ReferenceType
-import com.fasterxml.jackson.databind.type.TypeBindings
-import com.fasterxml.jackson.databind.type.TypeFactory
-import com.fasterxml.jackson.databind.type.TypeModifier
-import com.fasterxml.jackson.databind.util.AccessPattern
-import com.fasterxml.jackson.databind.util.NameTransformer
-import java.lang.reflect.Type
 
 public object OptionModule :
   SimpleModule(OptionModule::class.java.canonicalName, PackageVersion.VERSION) {
+  private val module: GenericTriStateModule<Option<*>> =
+    GenericTriStateModule(
+      serializationConfig =
+        GenericTriStateSerializationConfig(
+          isPresent = { it.isDefined() },
+          serializeValue = {
+            it.fold({ SerializedValue.AbsentOrNull }, { value -> SerializedValue.Value(value) })
+          }
+        ),
+      deserializationConfig =
+        GenericTriStateDeserializationConfig(
+          ifAbsent = { None },
+          ifNull = { None },
+          ifDefined = { it.some() }
+        )
+    )
 
   init {
-    addDeserializer(Option::class.java, OptionDeserializer())
+    addDeserializer(Option::class.java, module.deserializer)
   }
 
   override fun setupModule(context: SetupContext) {
     super.setupModule(context)
-    context.addSerializers(OptionSerializerResolver)
-    context.addTypeModifier(OptionTypeModifier)
+    context.addSerializers(module.serializerResolver)
+    context.addTypeModifier(module.typeModifier)
   }
 }
-
-public object OptionSerializerResolver : Serializers.Base() {
```

Comment:
All these public members need to have a deprecation cycle

Reply:
Thanks @i-walker that's a really good point. I'll defer the refactoring of option module as another issue entirely. I think several of these modules access modifier shouldn't have been public. I'll address that separately after this PR.